### PR TITLE
*: put the stats map to stats handle to make test more easier.

### PR DESCRIPTION
**Pull request by hanfei1991**
@shenli @coocood @zimulala @lamxTyler PTAL

### CODE REVIEW

**Comment by hanfei1991** [comment]
I bind stats handle with context, so every context can have its own stats, which may make test easier.

File: domain/domain.go [hunk 1]
```diff
@@ -407,15 +407,17 @@ func (do *Domain) PrivilegeHandle() *privileges.Handle {
 	return do.privHandle
 }
 
-// StatsHandle returns the statistic handle.
-func (do *Domain) StatsHandle() *statistics.Handle {
+// GetOrCreateStatsHandle returns the statistic handle. If the stats handle doesn't exist, it will create a new one.
```
Comment (lamxTyler) [inline domain/domain.go]: Just a little question here. Is it a good convention to explicit to describe how we get it in the function name? Or simply it doesn't matter? For example, `GetStatisticsTableCache` will create a pseudo one if it does not exist, do we have to say it in the function name?

Comment (hanfei1991) [inline domain/domain.go]: Well, I have no idea about this. @coocood @shenli What do you think of it ?

**Comment by lamxTyler** [comment]
What do you mean by saying that making the test easier? Every line that changed in the test is almost in the form of `GetStatisticsTableCache` to `GetStatsHandle(ctx).GetStatsTable`.

**Comment by hanfei1991** [comment]
1. different session can hold different histograms, we can simulate the case of multi nodes.
2. we can mock histograms for a query. It is thread safe when we run tests concurrently.

**Comment by lamxTyler** [comment]
@hanfei1991 Got it.

**Comment by hanfei1991** [comment]
@lamxTyler Currently we can only simulate a single node that crash after analyzing, and load it again.

File: domain/domain.go (hunk 1 above)
Comment (shenli) [inline domain/domain.go]: @lamxTyler I am agree with you. StatsHandle or GetStatsHandle  is better.

**Comment by hanfei1991** [comment]
@shenli @lamxTyler PTAL

File: statistics/statscache.go [hunk 2]
```diff
@@ -69,24 +74,17 @@ func (h *Handle) Update(is infoschema.InfoSchema) error {
...
-
-var statsTblCache = statsCache{cache: map[int64]*statsInfo{}}
-
-// GetStatisticsTableCache retrieves the statistics table from cache, and the cache will be updated by a goroutine.
-func GetStatisticsTableCache(tblInfo *model.TableInfo) *Table {
-	statsTblCache.m.RLock()
-	defer statsTblCache.m.RUnlock()
-	stats, ok := statsTblCache.cache[tblInfo.ID]
+// GetStatsTable retrieves the statistics table from cache, and the cache will be updated by a goroutine.
+func (h *Handle) GetStatsTable(tblInfo *model.TableInfo) *Table {
```
Comment (shenli) [inline statistics/statscache.go]: s/GetStatsTable/GetTableStats/

File: statistics/statscache.go [hunk 3]
```diff
@@ -99,17 +97,17 @@ func GetStatisticsTableCache(tblInfo *model.TableInfo) *Table {
 	return PseudoTable(tblInfo)
 }
 
-// SetStatisticsTableCache sets the statistics table cache.
-func SetStatisticsTableCache(id int64, statsTbl *Table, version uint64) {
-	statsTblCache.m.Lock()
-	defer statsTblCache.m.Unlock()
-	stats, ok := statsTblCache.cache[id]
+// SetStatsTable sets the statistics table cache.
+func (h *Handle) SetStatsTable(id int64, statsTbl *Table, version uint64) {
```
Comment (shenli) [inline statistics/statscache.go]: s/SetStatsTable/SetTableStats/

**Comment by lamxTyler** [comment]
LGTM

File: domain/domain.go [hunk 4]
```diff
@@ -407,15 +407,17 @@ func (do *Domain) PrivilegeHandle() *privileges.Handle {
 	return do.privHandle
 }
 
-// StatsHandle returns the statistic handle.
-func (do *Domain) StatsHandle() *statistics.Handle {
+// StatsHandle returns the statistic handle. If the stats handle doesn't exist, it will create a new one.
+func (do *Domain) StatsHandle(ctx context.Context) *statistics.Handle {
```
Comment (coocood) [inline domain/domain.go]: Why not init it in `NewDomain`?

**Comment by hanfei1991** [comment]
@coocood PTAL

File: statistics/table.go [hunk 5]
```diff
@@ -49,14 +49,14 @@ type Table struct {
 }
 
 // SaveToStorage saves stats table to storage.
-func (t *Table) SaveToStorage(ctx context.Context) error {
+func (t *Table) SaveToStorage(ctx context.Context, h *Handle) error {
```
Comment (coocood) [inline statistics/table.go]: Us handle as the method receiver is better.

**Pull request merged**